### ISSUE TITLE
improve require/import/export support in unminify

### ISSUE THREAD

**Issue by 0xdevalias** [j4k0xb/webcrack#30]
Was testing the new [v2.11.0](https://github.com/j4k0xb/webcrack/releases/tag/v2.11.0) web IDE update today and found some areas that unminify could be further improved.

This issue is about the require/import/export syntax after unminification.

You can get the minimised code that I am testing against here ([Ref](https://github.com/0xdevalias/chatgpt-source-watch/blob/main/orig/_next/static/chunks/167-121de668c4456907.js), that repo also has lots of other example code if you want to test further)

Loading that in the webcrack web IDE ([Ref](https://webcrack.netlify.app/)) with the following config:

![image](https://github.com/j4k0xb/webcrack/assets/753891/2d749c6b-9201-4742-afa7-cd1149d76c66)

In the unminified output, you can see an example of the current require/import/export in `180.js`:

_(**Note:** I also used this same file as my example for the JSX unminify in https://github.com/j4k0xb/webcrack/issues/10#issuecomment-1859476434, but the specifics in this issue are different, despite it using the same source)_

`180.js`
```js
require.d(exports, {
  Z: function () {
    return a;
  }
});
var r = require( /*webcrack:missing*/"./35250.js");
function a(e) {
  var t;
  var n = e.url;
  var a = e.size;
  var i = a === undefined ? 16 : a;
  var s = e.className;
  try {
    t = new URL(n);
  } catch (e) {
    console.error(e);
    return null;
  }
  return (0, r.jsx)("img", {
    src: `https://icons.duckduckgo.com/ip3/${t.hostname}.ico`,
    alt: "Favicon",
    width: i,
    height: i,
    className: s
  });
}
```

Contrasting this against the unminified output from `wakaru`'s web IDE ([Ref](https://wakaru.vercel.app/)) (which separates the 'unpack modules' and 'unminify' steps)

Unpacked `module-180.js` source:

```js
var r = require(35250);
function a(e) {
  var t;
  var n = e.url;
  var a = e.size;
  var i = void 0 === a ? 16 : a;
  var s = e.className;
  try {
    t = new URL(n);
  } catch (e) {
    return console.error(e), null;
  }
  return (0, r.jsx)("img", {
    src: "https://icons.duckduckgo.com/ip3/".concat(t.hostname, ".ico"),
    alt: "Favicon",
    width: i,
    height: i,
    className: s,
  });
}

module.exports = {
  Z: a,
};

```

Transformed (unminified) `module-180.js` source:

```js
const { jsx } = require(35250);

function a(e) {
  let t;

  const { url, size, className } = e;

  const i = size === undefined ? 16 : size;
  try {
    t = new URL(url);
  } catch (e) {
    console.error(e);
    return null;
  }
  return (
    <img
      src={`https://icons.duckduckgo.com/ip3/${t.hostname}.ico`}
      alt="Favicon"
      width={i}
      height={i}
      className={className}
    />
  );
}

export default {
  Z: a,
};
```

**Comment by j4k0xb** [comment]
Yeah it needs to get reworked at some point... the tricky part is detecting if a module is esm or not if it doesn't use `require.d` or similar (`10604.js`).
`require` should be converted to `import { jsx } from "./35250.js";` which will also solve #6 and partly #10

The default export is commonly called `Z`: https://github.com/swc-project/swc/blob/3550d0be26ebab7bcb7942e1a12a7d5fc3aedb17/crates/swc_ecma_minifier/tests/fixture/projects/next/archive-2/pages/hmr/.counter-c6a48518c4a56e4d9ce5/input.js#L10-L15

But it's still accessed like any other named export: https://github.com/swc-project/swc/blob/3550d0be26ebab7bcb7942e1a12a7d5fc3aedb17/crates/swc_ecma_minifier/tests/fixture/projects/next/archive-2/pages/hmr/.counter-c6a48518c4a56e4d9ce5/input.js#L230
so maybe `export { a as Z };` or `export function Z(e) {`

**Comment by 0xdevalias** [comment]
> the tricky part is detecting if a module is esm or not if it doesn't use `require.d` or similar

Forgive my naivety, but would it matter either way? Could it just be a user override setting of like 'always produce ESM', even if the original might not have been? Or would it introduce issues into the code that would make it not work/etc?

I guess based on this note ([Ref](https://github.com/pionxzh/wakaru/tree/main/packages/unminify#un-esm-unsafe)) it seems there may be incompatibilities if this was done.

**Comment by j4k0xb** [comment]
> Some packages require import * as name from 'package' instead of import name from 'package'. We cannot detect this automatically, so you might need to fix it manually.

And because you can't import esm from commonjs but that's impossible to validate with multiple chunks or other code we don't know about

Here's a wip version that converts top level require and some exports: https://deploy-preview-31--webcrack.netlify.app/
Output:
```jsx
import { jsx } from /*webcrack:missing*/"./35250.js";
export function Z(e) {
  var t;
  var n = e.url;
  var a = e.size;
  var i = a === undefined ? 16 : a;
  var s = e.className;
  try {
    t = new URL(n);
  } catch (e) {
    console.error(e);
    return null;
  }
  return <img src={`https://icons.duckduckgo.com/ip3/${t.hostname}.ico`} alt="Favicon" width={i} height={i} className={s} />;
}
```

**Comment by 0xdevalias** [comment]
> but that's impossible to validate with multiple chunks or other code we don't know about

@j4k0xb That makes sense for code we don't know about, but why is it impossible to validate for multiple chunks? Is that just because webcrack only operates on a single chunk at a time currently? ([Ref](https://deploy-preview-31--webcrack.netlify.app/docs/concepts/unpack.html))

**Comment by 0xdevalias** [comment]
> but that's impossible to validate with multiple chunks or other code we don't know about

@j4k0xb That makes sense for code we don't know about, but why is it impossible to validate for multiple chunks? Is that just because webcrack only operates on a single chunk at a time currently? ([Ref](https://deploy-preview-31--webcrack.netlify.app/docs/concepts/unpack.html))

> Here's a wip version that converts all top level requires and this export variation

> now the references will also be detected:

Nice! That's looking so much better/more useful already!

---

Using that same example code ([Ref](https://github.com/0xdevalias/chatgpt-source-watch/blob/main/orig/_next/static/chunks/167-121de668c4456907.js)), with the new preview version ([Ref](https://deploy-preview-31--webcrack.netlify.app/)), looking at `63390.js`, there are some 'top of file' `import`s, but then around line `108` there seem to be some more `import`s that aren't currently hoisted to the top of the file.

<details>
<summary>Details</summary>

Top of file imports:

```js
// 63390.js, lines 1-24
var r;
import { _ as _2 } from /*webcrack:missing*/"./39324.js";
import { _ as _3 } from /*webcrack:missing*/"./70216.js";
import { _ as _4 } from /*webcrack:missing*/"./4337.js";
import { jsxs, Fragment, jsx } from /*webcrack:missing*/"./35250.js";
import { Z as _Z } from /*webcrack:missing*/"./19841.js";
import { useRouter } from /*webcrack:missing*/"./68555.js";
import { useId, useState, useRef, useEffect, useCallback, forwardRef, Children, useMemo, memo } from /*webcrack:missing*/"./70079.js";
import { Z as _Z2 } from /*webcrack:missing*/"./34303.js";
import { Dd } from "./75179.js";
import { hz } from /*webcrack:missing*/"./64135.js";
import { FZ } from /*webcrack:missing*/"./88038.js";
import { tQ } from "./75527.js";
import { Qd } from "./36716.js";
import { wp } from /*webcrack:missing*/"./59710.js";
import { RR } from "./56244.js";
import { z as _z } from /*webcrack:missing*/"./66958.js";
import { Z as _Z3 } from "./30931.js";
import { Z as _Z4 } from "./87105.js";
import { _ as _5 } from /*webcrack:missing*/"./22830.js";
import { ZP as _ZP } from /*webcrack:missing*/"./2827.js";
import { WS } from /*webcrack:missing*/"./82841.js";
import { s6 } from /*webcrack:missing*/"./36218.js";
import { Os, uU } from "./69403.js";
```

Part way down, non-hoisted:

```js
// 63390.js, lines 108-115
import { _ } from /*webcrack:missing*/"./21722.js";
import { Jh } from /*webcrack:missing*/"./39889.js";
import { h } from /*webcrack:missing*/"./10642.js";
import { get } from /*webcrack:missing*/"./47635.js";
import { kP } from /*webcrack:missing*/"./32787.js";
import { ZP as _ZP2 } from /*webcrack:missing*/"./24274.js";
import { Z as _Z5 } from "./75515.js";
import { Z as _Z6 } from "./180.js";
```

</details>

With `require` and code with potential side effects, having them halfway down could change the logic; but if they are truly `import`s, then hoisting them to the top of the file should be fine I believe.

From memory, this web app is bundled with `swc`; I wonder if it has specific logic that knows (or is told) when an import/require halfway through a file isn't safe to be hoisted; and if so, whether we could leverage that knowledge to know when not to change a `require` to an `import`?

Looking at the same file/output in `wakaru` ([Ref](https://wakaru.vercel.app/file/63390)), it looks like it also doesn't hoist them, but it seems that it uses a mixture of both `import` and `require`, and these 'part way down' entries seem to be using `require`. I haven't deeply looked into the logic to validate why that is. (but see the 'Details' below, in the 'unminified' section, it seems that at least some of the `require`s for non-missing module files were converted to `import` and hoisted to the top of the file)

<details>
<summary>Details</summary>

## Top of File

**Source (unpacked)**

```js
// module-63390.js, lines 1-24
var r;
var a = require(39324);
var i = require(70216);
var s = require(4337);
var o = require(35250);
var l = require(19841);
var u = require(68555);
var d = require(70079);
var c = require(34303);
var f = require(75179);
var h = require(64135);
var g = require(88038);
var m = require(75527);
var p = require(36716);
var v = require(59710);
var x = require(56244);
var b = require(66958);
var y = require(30931);
var w = require(87105);
var j = require(22830);
var _ = require(2827);
var C = require(82841);
var M = require(36218);
var k = require(69403);
```

**Transformed (unminified)**

Note that `wakaru` doesn't currently explicitly annotate missing modules ([Ref](https://github.com/pionxzh/wakaru/issues/77)), so it could be that these 'top of file' `require`s that weren't converted to `import` is just because the modules are missing.

```js
// module-63390.js, lines 1-49
import f from "module-75179.js";
import m from "module-75527.js";
import p, { Qd } from "module-36716.js";
import x, { RR } from "module-56244.js";
import y from "module-30931.js";
import w from "module-87105.js";
import k from "module-69403.js";
import R from "module-75515.js";
import U from "module-180.js";
import Y, { dQ } from "module-77442.js";
import { v as v$0 } from "module-31721.js";
import ec, { Y8, m0 } from "module-5046.js";
import { Iy } from "module-25094.js";
let r;

const { _: _$5 } = require(39324);

const { _: _$4 } = require(70216);

const { _: _$0 } = require(4337);

const o = require(35250);

const { jsxs, jsx } = o;

const { Z: Z$0 } = require(19841);

const { useRouter } = require(68555);

const d = require(70079);

const { useId, useState, useRef, useEffect, useCallback, forwardRef, useMemo } =
  d;

const c = require(34303);

const { hz } = require(64135);

const g = require(88038);
const v = require(59710);
const b = require(66958);

const { _: _$1 } = require(22830);

const _ = require(2827);

const { WS } = require(82841);

const M = require(36218);
```

## Part way down

**Source (unpacked)**

```js
// module-63390.js, lines 138-146
var S = c.Z.div(N());
var I = require(21722);
var F = require(39889);
var E = require(10642);
var D = require(47635);
var L = require(32787);
var A = require(24274);
var R = require(75515);
var U = require(180);
```

**Transformed (unminified)**

```js
// module-63390.js, lines 170-182
var S = c.Z.div(N());

const { _: _$3 } = require(21722);

const { Jh } = require(39889);

const { h: h$0 } = require(10642);

const D = require(47635);

const { kP } = require(32787);

const A = require(24274);
```

Note that in the 'unpacked' for this section, there are `require`s that are no longer present in this 'unminified' section (`75515`, `180`); looking closer, it seems that these were converted to `import`s, and hoisted to the top of the file:

```js
// module-63390.js, lines 8-9
import R from "module-75515.js";
import U from "module-180.js";
```

</details>

**Comment by j4k0xb** [comment]
> but why is it impossible to validate for multiple chunks? Is that just because webcrack only operates on a single chunk at a time currently?

Yes, or rather a lot of work to do it "properly". Requires operating on *all* chunks and building an import/require graph like in https://github.com/pionxzh/wakaru/issues/73.
Maybe there's a better solution

> Looking at the same file/output in wakaru ([Ref](https://wakaru.vercel.app/file/63390)), it looks like it also doesn't hoist them, but it seems that it uses a mixture of both import and require, and these 'part way down' entries seem to be using require

I think it was a mistake to convert *all* requires, wakaru not touching the 'part way down' ones is more correct.

> if they are truly imports, then hoisting them to the top of the file should be fine I believe.

I doubt imports are ever moved down so it's probably a module with both import/require (which webpack can bundle).

**Comment by 0xdevalias** [comment]
> Yes, or rather a lot of work to do it "properly". Requires operating on all chunks and building an import/require graph like in https://github.com/pionxzh/wakaru/issues/73.

@j4k0xb _nods_, yeah, makes sense.

I wonder if even 'not proper' support would be useful to add, and how hard that would be. I've found `wakaru`'s current approach pretty useful even without the 'module graph', in just being able to unpack all the chunks and explore them more manually (though that's probably a seperate musing/usability thing and not super related to the topic here)

> Maybe there's a better solution

I haven't deeply looked into this, and not for ages, but at one stage I remember having a thought that the chunks specified the other chunks they depended on somewhere (as well as the individual module imports within it). 

In the code I was most exploring, theres the `_buildManifest.js` ([Ref](https://github.com/0xdevalias/chatgpt-source-watch/blob/main/unpacked/_next/static/%5BbuildHash%5D/_buildManifest.js)) and `webpack.js` ([Ref](https://github.com/0xdevalias/chatgpt-source-watch/blob/main/unpacked/_next/static/chunks/webpack.js)) chunks that seemed to detail some of the 'high level' of the chunk loading/dependencies/etc; though there was also the chunks loaded directly in the html as well

---

> I think it was a mistake to convert all requires, what wakaru does with not touching the 'part way down' ones is more correct.

@j4k0xb _nods_ makes sense. I'm not sure of the heuristics it uses to decide if they can be converted to `import` (and how 'smart' they are), but at least for some of the ones that weren't converted, I think that may have been due to them being for missing files.

---

> I doubt imports are ever moved down so it's probably a module with only `require` or both `import`/`require` (which webpack can bundle).

@j4k0xb Yeah, the 'import/require' part was what my original theory was.

**Comment by pionxzh** [comment]
> Looking at the same file/output in wakaru ([Ref](https://wakaru.vercel.app/file/63390)), it looks like it also doesn't hoist them, but it seems that it uses a mixture of both import and require, and these 'part way down' entries seem to be using require. I haven't deeply looked into the logic to validate why that is.

The current behavior of wakaru is to convert all top-level `require`s to `import`s, ignoring the risk of potentially changing the behavior. The reason why those `require`s in the sample are not converted is because they are missing chunks. wakaru also has a `hoist` flag to convert and hoist all nested requires to top-level `import`s, but there is currently no way for users to pass flags 😓  

I'm not sure if we can tell if a `require` should be converted to imports even with cross-module analysis.

**Comment by 0xdevalias** [comment]
> I'm not sure if we can tell if a `require` should be converted to imports even with cross-module analysis.

I haven't looked into this too deeply, but in cases where there is something similar to this, it should be safe to assume it's truly an ESM module yeah?

```js
Object.defineProperty(exports, "__esModule", { value: true });
```

Like this sort of thing:

> The modules are also converted to esm (more or less) if they're inside of a bundle and it removes `require.r`, `require.d`.
> What's still left is `require.j` and `module = require.nmd(module)` and `Object.defineProperty(exports, "__esModule", { value: true });`
> 
> _Originally posted by @j4k0xb in https://github.com/j4k0xb/webcrack/issues/33#issuecomment-1868384779_

**Comment by j4k0xb** [comment]
afaik `Object.defineProperty(exports, "__esModule", { value: true });` is inserted by tools like TypeScript and `__webpack_require__.r` by webpack, and nextjs does its own thing by not including any of this information 🥲

I looked at many of your samples and sometimes esm appears to be commonjs because of variable inlining or other build tools.
e.g.: <https://github.com/0xdevalias/chatgpt-source-watch/blob/main/orig/_next/static/chunks/167-121de668c4456907.js>  module 75527:
```js
var m = __webpack_require__(57311); // import ... from '57311', how webpack normally creates it
__webpack_require__(44675).env.INTERNAL_API_URL; // was probably `var n = __webpack_require__(44675); n.env.INTERNAL_API_URL`
```
or https://github.com/0xdevalias/chatgpt-source-watch/blob/main/orig/_next/static/chunks/293-defd068c38bd0c8d.js module 12285, temporary variables from typescript enums:
```js
__webpack_require__.d(exports, {
  R: function () {
    return V;
  },
});
var r;
var o;
var i;
var a;
var u = __webpack_require__(70079);
var c = __webpack_require__(62530);
var s = __webpack_require__(19430);
var l = __webpack_require__(9335);
var f = __webpack_require__(41800);
function p(t, e) {
  let [n, r] = (0, u.useState)(t);
  let o = (0, f.E)(t);
  (0, l.e)(() => r(o.current), [o, r, ...e]);
  return n;
}
var v = __webpack_require__(84325);
// ...
(r = P || {})[r.Open = 0] = "Open";
r[r.Closed = 1] = "Closed";
var P = r;
```

> And because you can't import esm from commonjs

did some more tests and apparently webpack is able to bundle it, even mixed commonjs/esm in a single module
so it should be fine to try to convert to esm when its obvious (without having to rely on analyzing other modules) and just keep the rest as commonjs

**Comment by j4k0xb** [comment]
#50 / https://deploy-preview-50--webcrack.netlify.app/ contains a rewrite of the whole unpacking/esm logic (very experimental)
currently it passes most of these test cases I mentioned earlier
https://github.com/j4k0xb/webcrack/blob/de98b60f2a7266bd6d955857dda107b450494d0d/packages/webcrack/src/unpack/test/imports.test.ts
https://github.com/j4k0xb/webcrack/blob/de98b60f2a7266bd6d955857dda107b450494d0d/packages/webcrack/src/unpack/test/exports.test.ts
